tests/mr/device/arena_mr_tests.cpp

```diff
@@ -533,6 +533,42 @@ TEST_F(ArenaTest, Defragment)  // NOLINT
   }());
 }
 
+TEST_F(ArenaTest, PerThreadAndStreamArenas)  // NOLINT
+{
+  auto const arena_size = superblock::minimum_size * 4;
+  arena_mr mr(rmm::mr::get_current_device_resource(), arena_size);
+  std::vector<std::thread> threads;
+  std::size_t num_threads{3};
+  auto view        = std::make_shared<rmm::cuda_stream_view>(rmm::cuda_stream_per_thread);
+  void* thread_ptr = mr.allocate(256, view->value());
+  threads.reserve(num_threads);
+  for (std::size_t i = 0; i < num_threads; ++i) {
+    threads.emplace_back(std::thread([&] {
+      cuda_stream stream{};
+      void* ptr = mr.allocate(32_KiB, stream);
+      mr.deallocate(ptr, 32_KiB, stream);
+    }));
+  }
+  for (auto& thread : threads) {
+    thread.join();
+  }
+  // The next allocation causes defrag, which causes all superblocks
+  // from the arenas allocated above to go back to global arena.
+  // Then we want to allocate all from global arena to make sure
+  // the superblock where the stream per thread allocated from is now
+  // owned by a stream arena instead of a thread arena or the global arena.
+  auto* ptr1 = mr.allocate(superblock::minimum_size);
+  auto* ptr2 = mr.allocate(superblock::minimum_size);
+  auto* ptr3 = mr.allocate(superblock::minimum_size);
```

Comment:
These can't be satisfied by the defragged existing superblock because that only has 1MiB - 256bytes space.

So now, the superblock that contains `thread_ptr` lives in the `set` maintained by the global arena.

Deallocating at this point was something that was already handled.